python/ray/serve/_private/deployment_state.py

```diff
@@ -1506,24 +1520,40 @@ def autoscale(
             current_num_ongoing_requests=current_num_ongoing_requests,
             current_handle_queued_queries=current_handle_queued_queries,
         )
-        if decision_num_replicas == self._target_state.num_replicas:
-            return
 
-        # Adjust the logged replica count based on `target_capacity`. In reality the
-        # autoscaling algorithm operates on the full replica count and the target
-        # will be adjusted in the in the main deployment update cycle, but this log
-        # statement needs to be comprehensible to the user.
-        adjusted_decision_num_replicas = self.get_capacity_adjusted_num_replicas(
-            decision_num_replicas, target_capacity
+        # Clip the replica count by capacity-adjusted bounds.
+        # TODO (shrekris-anyscale): this should logic should be pushed into the
+        # autoscaling_policy. Need to discuss what the right API would look like.
+        upper_bound = self.get_capacity_adjusted_num_replicas(
+            autoscaling_policy.config.max_replicas, target_capacity
         )
+        if (
+            target_capacity_scale_direction == TargetCapacityScaleDirection.UP
+            and autoscaling_policy.config.initial_replicas is not None
+        ):
+            lower_bound = self.get_capacity_adjusted_num_replicas(
+                autoscaling_policy.config.initial_replicas, target_capacity
+            )
+        else:
```

Comment:
A bit nit, but I feel when `target_capacity_scale_direction` is None, lower bound should fallback to the using `initial_replicas`?

Reply:
Not quite– if `target_capacity_scale_direction` is `None`, then Serve has finished the upgrade. At that point, we can safely ignore `initial_replicas` and use `min_replicas`/`max_replicas` as the bounds.